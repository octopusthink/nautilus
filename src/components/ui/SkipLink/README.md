A `<SkipLink>` is a hidden internal link at the beginning of a page that provides a way for keyboard users to bypass navigation element and skip directly to the main content.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Paragraph>Click or tab here, then press `tab` to move focus to the `SkipLink` component.</Paragraph>

  <SkipLink toId="/Function/UI Components/SkipLink" />

  <Paragraph>Pressing tab again will hide the `SkipLink`.</Paragraph>
</React.Fragment>
```

## Usage

Your app/site should have a single `<SkipLink>`, located early in your content so it is easily activated by a keyboard user when they start to tab through content.

### Use this component for…

- Providing an easy way for users to skip to the main content of your page.
- Any page that includes page navigation. Without a skip link, keyboard users will find it incredibly difficult to navigate your site.

### Don't use this component for…

- Providing multiple skip links to different sections of your page. Divide your content up with `<Heading>`s or consider separating your content into separate pages instead.

## Appearance

`SkipLink`s are visually hidden until they have focus. When focused they should be visible and appear at the top of the page.

## Behaviour

The skip link component is visually hidden until activated by a keyboard press. When activated, the skip link appears and gains focus. If the user then presses the enter key, they will automatically be taken to the main content container.

By default, a `SkipLink` moves focus to a container with an ID of `#content`, but this can be changed if needed. For example, if your main content container has an ID of `main`, you should specify this using the `toId` prop:

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Paragraph>Tab or click here, then press tab to view the skip link.</Paragraph>
  <SkipLink toId="main" />
</React.Fragment>
```

## Voice & tone

You can customise the text inside a skip link if needed for translations or otherwise. The text for the skip link should be clear, concise, and clearly describe what the link does.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Paragraph>Tab or click here, then press tab to view the skip link.</Paragraph>
  <SkipLink>Jump to content</SkipLink>
</React.Fragment>
```

## Accessibility

Skip links will be used by users already familiar with your site's navigation or those who wish to bypass the navigation and immediately jump to the main content. This is analogous to a user who scrolls past a site's header right to content without reading it).

A skip link should be the first element that is focused when the user presses the `tab` key.

Skip links are a very useful accessibility feature, but they should not be an excuse to create overly cumbersome navigation. Keyboard users will still likely need to interact with your navigation elements and other elements that are skipped over by your `SkipLink` component.
