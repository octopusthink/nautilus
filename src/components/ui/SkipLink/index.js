import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React from 'react';

import Link from 'components/ui/Link';
import { VisuallyHiddenStyles, VisuallyHiddenRevealStyles } from 'components/ui/VisuallyHidden';
import { interfaceUI, focusStyle, toUnits } from 'styles';
import { useTheme } from 'themes';

export const SkipLink = (props) => {
  const { children, toId, unstyled, ...otherProps } = props;

  const theme = useTheme();

  return (
    <Link
      as="a"
      css={
        unstyled
          ? undefined
          : css`
              /*
                We should hide the SkipLink by default, only showing it when its
                been focused by the user.
              */
              ${VisuallyHiddenStyles}
              transition: none;

              &:focus {
                ${VisuallyHiddenRevealStyles}
                ${interfaceUI.medium(theme)};
                ${focusStyle.outline(theme)};
                ${focusStyle.text(theme)};
                background: ${theme.colors.neutral.white};
                border: 0;
                box-sizing: border-box;
                display: block;
                left: 0;
                padding: ${toUnits(theme.spacing.padding.medium)};
                position: absolute;
                text-decoration: underline;
                text-align: center;
                top: 0;
                width: 100%;
                z-index: 100;
              }
            `
      }
      {...otherProps}
      href={`#${toId}`}
    >
      {children}
    </Link>
  );
};

SkipLink.defaultProps = {
  children: 'Skip to main content',
  toId: 'content',
  unstyled: false,
};

SkipLink.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** The HTML ID of the main "content" element on the page. This is what the skip link will take the user to when used. */
  toId: PropTypes.string,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = SkipLink;

export default SkipLink;
