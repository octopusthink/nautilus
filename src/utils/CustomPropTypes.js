import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
