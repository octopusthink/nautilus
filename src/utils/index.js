export { default as CustomPropTypes } from './CustomPropTypes';
