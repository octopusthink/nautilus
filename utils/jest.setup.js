// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
