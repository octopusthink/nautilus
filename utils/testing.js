// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
