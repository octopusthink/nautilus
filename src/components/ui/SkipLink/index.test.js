import React, { createRef } from 'react';

import { renderWithRouter } from 'utils/testing';

import SkipLink from '.';

describe('SkipLink', () => {
  it('should render an <a> tag', () => {
    const { container } = renderWithRouter(<SkipLink>Hello</SkipLink>);

    expect(container.firstChild.tagName).toEqual('A');
  });

  it('should output its children', () => {
    const { getByTestId } = renderWithRouter(
      <SkipLink>
        <span data-testid="child">Skip to content</span>
      </SkipLink>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = renderWithRouter(
      <SkipLink className="custom-class" data-testid="myText">
        hello
      </SkipLink>,
    );

    expect(getByTestId('myText').classList).toContain('custom-class');
  });

  it('should forward refs', () => {
    const ref = createRef();

    renderWithRouter(<SkipLink ref={ref}>Homepage</SkipLink>);

    expect(ref.current).not.toBeNull();
    expect(ref.current.tagName).toEqual('A');
  });
});
