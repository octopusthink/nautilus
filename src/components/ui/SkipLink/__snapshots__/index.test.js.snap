// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipLink should contain styles, including focus styles that reveal the link 1`] = `
.emotion-0 {
  color: #cd2f83;
  box-shadow: 0 2px #f5bbda;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  border: 0;
  -webkit-clip-path: rect(0 0 0 0);
  clip-path: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  -webkit-transition: none;
  transition: none;
}

.emotion-0:hover {
  box-shadow: 0 2px #a0dee3;
  color: #0097a2;
}

.emotion-0:focus {
  box-shadow: 0 0 0 0.2rem #f5bbda;
  outline: 0;
  color: #cd2f83;
}

.emotion-0:focus {
  -webkit-clip-path: none;
  clip-path: none;
  height: auto;
  margin: 0;
  overflow: auto;
  position: static;
  white-space: normal;
  width: auto;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 1.3333333333333333;
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
  color: #cd2f83;
  background: #ffffff;
  border: 0;
  box-sizing: border-box;
  display: block;
  left: 0;
  padding: 1.2rem;
  position: absolute;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-align: center;
  top: 0;
  width: 100%;
  z-index: 100;
}

<div>
  <a
    class="emotion-0"
    href="#content"
  >
    Skip to main content
  </a>
</div>
`;

exports[`SkipLink should not render styles when unstyled prop is set 1`] = `
<div>
  <a
    class="emotion-0"
    href="#content"
  >
    Skip to main content
  </a>
</div>
`;
